File: frontend/src/app/services/bookService/book.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Book } from 'src/app/models/bookModel/book';

@Injectable({
  providedIn: 'root',
})
export class BookService {
  private apiUrl = 'http://localhost:9292/books'; // Change this to match your backend URL

  constructor(private http: HttpClient) {}

  getBookById(isbn: number): Observable<Book> {
    return this.http.get<Book>(`${this.apiUrl}/${isbn}`);
  }

  addBook(book: Book): Observable<Book> {
    return this.http.post<Book>(`${this.apiUrl}/addBook`, book);
  }

  getAllBooks(): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.apiUrl}`);
  }

  searchBooksByIsbn(isbn: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.apiUrl}/search/isbn/${isbn}`);
  }

  searchBooksByTitle(title: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.apiUrl}/search/title/${title}`);
  }

  searchBooksByAuthor(author: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.apiUrl}/search/author/${author}`);
  }

  searchBooksByGenre(genre: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.apiUrl}/search/genre/${genre}`);
  }

  searchBooksByPublicationYear(publicationYear: number): Observable<Book[]> {
    return this.http.get<Book[]>(
      `${this.apiUrl}/search/publicationYear/${publicationYear}`
    );
  }
}

File: backend/src/main/java/com/fdb/backend/Services/BookService.java
package com.fdb.backend.Services;

import com.fdb.backend.Entities.Book;
import com.fdb.backend.Repositories.BookRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class BookService {

    @Autowired
    private BookRepository bookRepository;

    public List<Book> getAllBooks() {
        return bookRepository.findAll();
    }

    public Book getBookById(Long id) {
        return bookRepository.findById(id).orElse(null);
    }

    public Book addBook(Book book) {
        return bookRepository.save(book);
    }

    public void updateBook(Book book) {
        bookRepository.save(book);
    }

    public List<Book> searchBooksByIsbn(Long isbn) {
        // Implement search by ISBN logic using bookRepository
        return bookRepository.findByIsbn(isbn);
    }

    public List<Book> searchBooksByTitle(String title) {
        // Implement search by title logic using bookRepository
        return bookRepository.findByTitleContainingIgnoreCase(title);
    }

    public List<Book> searchBooksByAuthor(String author) {
        // Implement search by author logic using bookRepository
        return bookRepository.findByAuthorContainingIgnoreCase(author);
    }

    public List<Book> searchBooksByGenre(String genre) {
        // Implement search by genre logic using bookRepository
        return bookRepository.findByGenreContainingIgnoreCase(genre);
    }

    public List<Book> searchBooksByPublicationYear(Integer publicationYear) {
        // Implement search by publication year logic using bookRepository
        return bookRepository.findByPublicationYear(publicationYear);
    }

}
